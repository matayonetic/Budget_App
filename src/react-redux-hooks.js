import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useDispatch = () => originalUseDispatch();
export const useSelector = (state) => originalUseSelector(state);
