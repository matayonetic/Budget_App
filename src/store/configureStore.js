import { legacy_createStore as createStore, combineReducers } from "redux";

// Reducers
import expensesReducer from "../reducers/expensesReducer";
import filtersReducer from "../reducers/filtersReducer";

// Store
const configureStore = () => {
  const store = createStore(
    combineReducers({
      expenses: expensesReducer,
      filters: filtersReducer,
    }),
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );
  return store;
};

export { configureStore as default };
