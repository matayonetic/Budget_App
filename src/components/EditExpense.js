import React from "react";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expensesGen";
import { useNavigate, useParams } from "react-router-dom";
import { useSelector, useDispatch } from "../react-redux-hooks";

// Edit Expense Component
const EditExpense = () => {
  //
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  // Expense
  const expense = useSelector((state) =>
    state.expenses.find((expense) => expense.id === id)
  );

  // Edit Expense
  const editMyExpense = (expense) => {
    dispatch(startEditExpense(id, expense));
    navigate("/dashboard");
  };

  // Remove Expense
  const removeMyExpense = () => {
    dispatch(startRemoveExpense({ id }));
    navigate("/dashboard");
  };

  // Render
  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h2 className="page-header__title">Edit Expense</h2>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm expense={expense} formData={editMyExpense} />
        <button className="button button__secondary" onClick={removeMyExpense}>
          Remove Expense
        </button>
      </div>
    </div>
  );
};

export default EditExpense;
