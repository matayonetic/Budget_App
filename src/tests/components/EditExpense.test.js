import React from "react";
import EditExpense from "../../components/EditExpense";
import { shallow } from "enzyme";
import * as ReactReduxHooks from "../../react-redux-hooks";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import expenses from "../fixtures/expenses";

// Mock Nav & Params
const mockNavigate = jest.fn();
const mockParams = jest.fn();
jest.mock("react-router-dom", () => {
  return {
    useNavigate: () => mockNavigate,
    useParams: () => mockParams
  };
});

let store, wrapper, mockDispatch;

// Before each test
beforeEach(() => {
  //
  // Mock Store
  store = configureStore([thunk])(expenses[0]);

  // Mock Selector
  jest
    .spyOn(ReactReduxHooks, "useSelector")
    .mockImplementation(() => store.getState());

  // Mock Dispatch
  mockDispatch = jest.fn();
  jest.spyOn(ReactReduxHooks, "useDispatch").mockReturnValue(mockDispatch);
  // .mockImplementation(() => store.dispatch);
});

// Test Snapshot
test("Should render editExpense page", () => {
  wrapper = shallow(<EditExpense />);
  expect(wrapper).toMatchSnapshot();
});

// Testing formData prop
test("Should handle editMyExpense", () => {
  wrapper = shallow(<EditExpense />);
  wrapper.find("ExpenseForm").prop("formData")(expenses[0]);
  expect(mockNavigate).toHaveBeenLastCalledWith("/");  
});

// Testing removeExpense Handler
test("Should handle removeExpense", () => {
  wrapper = shallow(<EditExpense />);
  wrapper.find("button").simulate("click");
  expect(mockNavigate).toHaveBeenLastCalledWith("/");  
});
