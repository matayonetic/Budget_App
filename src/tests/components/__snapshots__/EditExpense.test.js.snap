// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render editExpense page 1`] = `
<div>
  <h2>
    Edit Expense
  </h2>
  <ExpenseForm
    expense={
      Object {
        "amount": 1,
        "createdAt": -86400000,
        "description": "Food",
        "id": "1",
        "note": "a",
      }
    }
    formData={[Function]}
  />
  <p>
    <button
      onClick={[Function]}
    >
      Remove
    </button>
  </p>
</div>
`;
